File: YLCategory-Swift-MacOS/NSResponder/NSResponder+category.m
//
//  NSResponder+category.m
//  YLCategory-MacOS
//
//  Created by 魏宇龙 on 2023/5/8.
//

#import "NSResponder+category.h"
#import <objc/runtime.h>

#ifndef kAppIsDarkTheme
// 判断当前app是否是深色模式
#define kAppIsDarkTheme                    ([NSApp.effectiveAppearance bestMatchFromAppearancesWithNames:@[NSAppearanceNameDarkAqua, NSAppearanceNameAqua]] == NSAppearanceNameDarkAqua)
#endif

#ifndef kSystemIsDarkTheme
// 系统当前是否是深色模式
#define kSystemIsDarkTheme          \
^ BOOL{                             \
    NSDictionary *dict = [[NSUserDefaults standardUserDefaults] persistentDomainForName:NSGlobalDomain];    \
    id style = [dict objectForKey:@"AppleInterfaceStyle"];  \
    return style && [style isKindOfClass:[NSString class]] && NSOrderedSame == [style caseInsensitiveCompare:@"dark"];  \
}()
#endif

@implementation NSResponder (category)

- (YLThemeChangedHandler)systemThemeChangedHandler {
    return objc_getAssociatedObject(self, _cmd);
}

- (void)setSystemThemeChangedHandler:(YLThemeChangedHandler)systemThemeChangedHandler {
    objc_setAssociatedObject(self, @selector(systemThemeChangedHandler), systemThemeChangedHandler, OBJC_ASSOCIATION_COPY_NONATOMIC);
    if(systemThemeChangedHandler) {
        [[NSDistributedNotificationCenter defaultCenter] addObserver:self selector:@selector(yl_systemThemeChanged) name:@"AppleInterfaceThemeChangedNotification" object:nil];
    } else {
        [[NSDistributedNotificationCenter defaultCenter] removeObserver:self name:@"AppleInterfaceThemeChangedNotification" object:nil];
    }
}

- (void)yl_systemThemeChanged {
    if (self.systemThemeChangedHandler) {
        self.systemThemeChangedHandler(self, kSystemIsDarkTheme);
    }
}

- (YLThemeChangedHandler)appThemeChangedHandler {
    return objc_getAssociatedObject(self, _cmd);
}

- (void)setAppThemeChangedHandler:(YLThemeChangedHandler)appThemeChangedHandler {
    objc_setAssociatedObject(self, @selector(appThemeChangedHandler), appThemeChangedHandler, OBJC_ASSOCIATION_COPY_NONATOMIC);
    if(appThemeChangedHandler) {
        [NSApp addObserver:self forKeyPath:@"effectiveAppearance" options:NSKeyValueObservingOptionNew context:nil];
    } else {
        [NSApp removeObserver:self forKeyPath:@"effectiveAppearance" context:nil];
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    if([keyPath isEqualToString:@"effectiveAppearance"]) {
        if(self.appThemeChangedHandler) {
            self.appThemeChangedHandler(self, kAppIsDarkTheme);
        }
    }
}

- (void)dealloc {
    [[NSDistributedNotificationCenter defaultCenter] removeObserver:self name:@"AppleInterfaceThemeChangedNotification" object:nil];
    if(self.appThemeChangedHandler) {
        [NSApp removeObserver:self forKeyPath:@"effectiveAppearance" context:nil];
    }
}

@end

File: YLCategory-Swift-MacOS/NSResponder/NSResponder+Category.swift
//
//  NSResponder+Category.swift
//  YLCategory-Swift-MacOS
//
//  Created by 魏宇龙 on 2025/2/19.
//

import AppKit

fileprivate var SystemThemeChangedHandlerKey = false
fileprivate var AppThemeChangedHandlerKey = false
fileprivate var AppThemeObserverKey = false

public extension NSResponder {
    
    // MARK: - 监听系统亮色｜暗色切换 （ ⚠️ 在不需要的时候，设置为nil）
    var systemThemeChangedHandler: ((NSResponder, Bool) -> Void)? {
        get { objc_getAssociatedObject(self, &SystemThemeChangedHandlerKey) as? (NSResponder, Bool) -> Void }
        set {
            objc_setAssociatedObject(self, &SystemThemeChangedHandlerKey, newValue, .OBJC_ASSOCIATION_COPY_NONATOMIC)
            if newValue != nil {
                DistributedNotificationCenter.default().addObserver(self, selector: #selector(systemThemeChangedNotification), name: NSNotification.Name("AppleInterfaceThemeChangedNotification"), object: nil)
                
            } else {
                DistributedNotificationCenter.default().removeObserver(self, name: NSNotification.Name("AppleInterfaceThemeChangedNotification"), object: nil)
            }
        }
    }
    
    @objc private func systemThemeChangedNotification(){
        systemThemeChangedHandler?(self, systemIsDarkTheme)
    }
    
    /// 系统是暗色模式
    private var systemIsDarkTheme: Bool {
        let info = UserDefaults.standard.persistentDomain(forName: UserDefaults.globalDomain)
        if let style = info?["AppleInterfaceStyle"] as? String {
            return style.caseInsensitiveCompare("dark") == .orderedSame
        }
        return false
    }
    
    // MARK: - 监听app 亮色｜暗色切换  (⚠️ 在不需要的时候，设置为nil）
    var appThemeChangedHandler: ((NSResponder, Bool) -> Void)? {
        get { objc_getAssociatedObject(self, &AppThemeChangedHandlerKey) as? (NSResponder, Bool) -> Void }
        set {
            objc_setAssociatedObject(self, &AppThemeChangedHandlerKey, newValue, .OBJC_ASSOCIATION_COPY_NONATOMIC)
            if newValue != nil {
                if !isAppObserverAdded {
                    NSApp.addObserver(self, forKeyPath: "effectiveAppearance", options: [.new], context: nil)
                    isAppObserverAdded = true
                }
            } else {
                if isAppObserverAdded {
                    NSApp.removeObserver(self, forKeyPath: "effectiveAppearance")
                    isAppObserverAdded = false
                }
            }
        }
    }
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if keyPath == "effectiveAppearance" {
            appThemeChangedHandler?(self, appIsDarkTheme)
        } else {
            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
        }
    }
    
    /// 是否已经设置监听app主题变化
    private var isAppObserverAdded: Bool {
        get { objc_getAssociatedObject(self, &AppThemeObserverKey) as? Bool ?? false }
        set { objc_setAssociatedObject(self, &AppThemeObserverKey, newValue, .OBJC_ASSOCIATION_ASSIGN) }
    }
    
    /// app是暗色模式
    private var appIsDarkTheme: Bool { NSApp.effectiveAppearance.bestMatch(from: [.darkAqua, .aqua]) == .darkAqua }
}
